### ISSUE TITLE
[WIP] Clarify the motivation behind the `flush` option

### ISSUE THREAD

**Issue by schomatis** [ipfs/go-mfs#37]
With various names throughout the code there's an option being provided by the current API with a vague promise of a "flush" operation. That is, when opening an MFS file there is a (externally called, e.g., in `go-ipfs`) `flush` argument (internally called `sync)` provided to the user but there isn't an exact description of what's implied by that and it seems to be used with different motivations in different use cases.

Internally, `flush`/`sync` means when closing a file descriptor (or, less common, flushing a file directly), the changes will be propagated upwards all the way to the root: the parent of the file will update its entry, the parent's parent will update its entry to point to the new directory that points to the new content of the file and so on. If unset only the parent's entry will be updated, which in itself raises a suspicious use case because it's not clear in which case the user would want that since normally the entire MFS is referenced by the hash of the root, and if the root isn't updated and only the first parent is, it would seem those changes would be lost. As an example see https://github.com/ipfs/go-mfs/issues/28 (and related).

In `go-ipfs` (the main consumer of MFS) the typical use of `flush` is to set it only when the file is being opened with the write mode (and leave it unset in read-only) which seems like an unnecessary distinction since the file descriptor already has a mechanism to detect modifications (through the internal `hasChanges` variable) to avoid unnecessary updates.

Also, the most important MFS-related command, `ipfs files`, sets the `flush` as a global and undocumented option which has the effect of leaving many MFS-write operations without effect (see the previously cited issue).

With all this in mind it seems like we should rethink in which case do we want to perform a local update of only the first parent and provide an option specifically for that and deprecate the current usage of `flush` (generating a full update by default).

**Comment by Stebalien** [comment]
So, *locally* MFS is usually accessed through the logical root, not the root hash. That means that, if everything is working correctly, non-flushed changes should still appear in commands like `ipfs files ls /...` and/or `ipfs files cat /...`.

The idea behind the flush flag is that flushing on every write is *extremely* slow as we need to compute multiple sha256 hashes on every flush. `--flush=false` allows us to defer this to some later time (finally flushing using `ipfs files flush`).

However, we *should* seriously consider the *semantics* of flush. For example, we might want to flush whenever a user stats a file (to find it's CID). On the other hand, that could have serious performance implications (at least once we start using MFS seriously).

**Comment by schomatis** [comment]
> The idea behind the flush flag is that flushing on every write is _extremely_ slow as we need to compute multiple sha256 hashes on every flush. `--flush=false` allows us to defer this to some later time (finally flushing using `ipfs files flush`).

Thanks for the clarification, this is something that may not be obvious to a new user (like me), I'll try to add it to the documentation.

For future references I'm leaving here the `loadFilesRoot` function which helped understand this much better, now I see that the publish function (I've been wrestling with) can be something simple like a `Datastore().Put("/local/filesroot")` call (it can also involve more complex operations in the IPNS case but this is a nice example):

https://github.com/ipfs/go-ipfs/blob/d0c27272ad84e8fa2b095b319f033a968d3b9351/core/core.go#L827

The error in my reasoning before was my implicit assumption of the problem that if I didn't do a full update in the moment while I still had the reference to say a newly created file I would later be unable to retrieve it in a final update of the MFS root (something like "either update it all now or lose it forever") but I hadn't been paying enough attention to what I thought it only served as a simple performance cache in the MFS directory:

https://github.com/ipfs/go-mfs/blob/2c1b835ceb3a181336651f2285dc4a13b9266c32/dir.go#L32-L33

(which now I see that should be thought of as something more since it has information not available in the UnixFS nodes).

These maps allow to keep a reference to a new/modified entry without needing a full flush every time, the final `Close` of the MFS root will trigger the `sync` function across all the entries in the MFS guaranteeing that at the end we have the final state of the filesystem:

https://github.com/ipfs/go-mfs/blob/2c1b835ceb3a181336651f2285dc4a13b9266c32/dir.go#L432-L443

(this explains why killing the daemon loses this state since we are not giving it time to correctly shut down while the offline case keeps working, see https://github.com/ipfs/go-ipfs/issues/5738#issuecomment-435944273).

**Comment by djdv** [comment]
I recently got tripped up on this.
After modifying a root, I was relying on its republisher to return before `root.Flush()` returned, as is the case if you where to call `mfs.FlushPath(root, "/")`. 
![mfs flush thing 01](https://user-images.githubusercontent.com/13862850/53188212-e2716f80-35d2-11e9-9b46-b5ebb6840d13.png)

Since that's not how it works, my mind initially went to renaming `root.Flush()` into `root.Repub()` (or whatever), and having `Flush()` call `Repub()` + `WaitPub()`, simillar to how `FlushPath` works.
Likewise for operations like `mfs.Mkdir`, splitting the flush option into 2 for whatever is desired. Blocking until replushing is finished when the flush option is provided.

I lack comprehension of the MFS internals though so this is not a suggestion, just noting it.

**Comment by aschmahmann** [comment]
This repository is no longer maintained and has been copied over to [Boxo](https://github.com/ipfs/boxo).  In an effort to avoid noise and crippling in the Boxo repo from the weight of issues of the past, we are closing most issues and PRs in this repo.  Please feel free to open a new issue in Boxo (and reference this issue) if resolving this issue is still critical for unblocking or improving your usecase.

You can learn more in the [FAQs for the Boxo repo copying/consolidation effort](https://github.com/ipfs/boxo/wiki/Copied-or-Migrated-Repos-FAQ).
